scalalib/src/mill/javalib/android/AndroidAppModule.scala

```diff
@@ -47,13 +49,32 @@ object AndroidLintReportFormat extends Enumeration {
 @mill.api.experimental
 trait AndroidAppModule extends JavaModule {
 
+  final def root: os.Path = super.millSourcePath
+  private def src: os.Path = root / "src"
+  override def millSourcePath: os.Path = src / "main"
+
+  override def sources: T[Seq[PathRef]] = Task.Sources(millSourcePath / "java")
+
+  override def repositoriesTask: Task[Seq[Repository]] = Task.Anon {
+    super.repositoriesTask() ++
+      Seq(MavenRepository("https://maven.google.com"))
+  }
+
+  /* TODO this is a temporary hack to exclude any jvm only tagged dependencies
+   * which may conflict with android dependencies. For example, the
+   * kotlin coroutines which are provided by both kotlin-core and
+   * kotlin-core-jvm . See also https://github.com/com-lihaoyi/mill/issues/3867
+   */
```

Comment:
What is the exact issue here? Duplicate class definition?

Reply:
yes, I am planning to look into consolidating the assembly functionality later (experimenting with a "fat jar", extracting that and see if d8 picks everything up), so I should be addressing it in a follow-up PR

Reply:
This is pretty strange, I didn't face that in my PR which also has coroutines in the dependency graph. Are you able to post resolved dependency graph you have without this override?

Reply:
have you managed to run the integration tests? this happens as a transitive dependency from androidx test something, let me change it and give you the exact error

Reply:
No, I didn't pull this PR to run it yet. Also FYI there is the following issue in Mill (well, in Coursier), because Gradle under the hood relies on the `.module` files rather than `.pom` for the dependency graph resolution https://github.com/com-lihaoyi/mill/issues/3993.

Reply:
the error is
```
[3074]   Caused by: com.android.tools.r8.internal.g: Type kotlinx.coroutines.StandaloneCoroutine is defined multiple times: /home/vnicolaou/.cache/coursier/v1/https/repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-core-jvm/1.7.1/kotlinx-coroutines-core-jvm-1.7.1.jar:kotlinx/coroutines/StandaloneCoroutine.class, /home/vnicolaou/.cache/coursier/v1/https/repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-core/1.3.4/kotlinx-coroutines-core-1.3.4.jar:kotlinx/coroutines/StandaloneCoroutine.class
```

and the relevant section of ivy dependency tree

```
├─ androidx.test:core:1.6.1
├─ androidx.concurrent:concurrent-futures-ktx:1.1.0
      └─ org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.4
      └─ org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.7.1
```

Reply:
I've read the thread in #3993 and I suspect you are right about the module. I also suspect that gradle or the android kotlin plugin (?) removes any dependencies that are not androidJvm
`"org.jetbrains.kotlin.platform.type": "androidJvm"`

as here
https://repo1.maven.org/maven2/org/jetbrains/kotlinx/kotlinx-coroutines-core-jvm/1.10.1/kotlinx-coroutines-core-jvm-1.10.1.module

the platform type is jvm

Reply:
> ├─ androidx.concurrent:concurrent-futures-ktx:1.1.0
      └─ org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.4
      └─ org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.7.1

How `kotlinx-coroutines-core-jvm` occurs in this graph? https://mvnrepository.com/artifact/androidx.concurrent/concurrent-futures-ktx/1.1.0 has a dependency on `org.jetbrains.kotlinx:kotlinx-coroutines-core`.

The problem here definitely is not in the presence of the `-jvm` suffix artifacts in the graph, but in the graph resolution itself.

Reply:
sorry, made an indentation mistake, it comes from `androidx.test:core:1.6.1`

the issue is not the suffix, it's just happens that filtering them out works, but I think the proper implementation would be to look into the .module files in dependency resolution, and filter by platform type

Reply:
I would propose to remove this from `AndroidAppModule`, because it may cause the issues further, when some dependency with `-jvm` suffix _must_ be in the resolved graph.

Instead, you can manage dependencies resolution in the build script of the example, adding TODO marker like here https://github.com/com-lihaoyi/mill/pull/4188/files#diff-5f6fc77ef82493c72b1011020468273b24df3b3454dd7dba6b0c266a91dd2e02R42-R57.

Reply:
tbh, I'm not happy with either, but I added the exclusion to the dependencies and the TODO for now.

Let me know if you are happy with this

Reply:
yes, looks better